{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Task-1.-Compiling-Ebola-Data\"><span class=\"toc-item-num\">Task 1.&nbsp;&nbsp;</span>Compiling Ebola Data</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-2.-RNA-Sequences\"><span class=\"toc-item-num\">Task 2.&nbsp;&nbsp;</span>RNA Sequences</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-3.-Class-War-in-Titanic\"><span class=\"toc-item-num\">Task 3.&nbsp;&nbsp;</span>Class War in Titanic</a></div></p>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This is a work version "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "from datetime import datetime\n",
    "from datetime import date, time\n",
    "from dateutil.parser import parse\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "pd.options.mode.chained_assignment = None\n",
    "\n",
    "DATA_FOLDER = \"/Users/karunya/Documents/EPFL/ADA/ADA2017-Tutorials/Intro_to_Pandas/Data\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1. Compiling Ebola Data\n",
    "\n",
    "The `DATA_FOLDER/ebola` folder contains summarized reports of Ebola cases from three countries (Guinea, Liberia and Sierra Leone) during the recent outbreak of the disease in West Africa. For each country, there are daily reports that contain various information about the outbreak in several cities in each country.\n",
    "\n",
    "Use pandas to import these data files into a single `Dataframe`.\n",
    "Using this `DataFrame`, calculate for *each country*, the *daily average* per year of *new cases* and *deaths*.\n",
    "Make sure you handle all the different expressions for *new cases* and *deaths* that are used in the reports."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Create 3 Dataframes (one for each country)\n",
    "filenames = glob.glob(DATA_FOLDER +\"/ebola/guinea_data/\"+\"*.csv\")\n",
    "get_dfs = (pd.read_csv(f) for f in filenames)\n",
    "guinea = pd.concat(get_dfs, ignore_index=True)\n",
    "guinea['Country']='Guinea'\n",
    "\n",
    "filenames = glob.glob(DATA_FOLDER +\"/ebola/liberia_data/\"+\"*.csv\")\n",
    "get_dfs = (pd.read_csv(f) for f in filenames)\n",
    "liberia = pd.concat(get_dfs, ignore_index=True)\n",
    "liberia['Country'] ='Liberia'\n",
    "\n",
    "filenames = glob.glob(DATA_FOLDER +\"/ebola/sl_data/\"+\"*.csv\")\n",
    "get_dfs = (pd.read_csv(f) for f in filenames)\n",
    "sl = pd.concat(get_dfs, ignore_index=True)\n",
    "sl['Country'] ='Sierra Leone'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Homogenize Column Names\n",
    "guinea.rename(columns={'Description': 'Variable', 'Totals':'National'}, inplace=True)\n",
    "sl.rename(columns={'date': 'Date', 'variable':'Variable'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Merge Tables\n",
    "liberia_table = liberia[['Country','Date','Variable','National']]\n",
    "guinea_table = guinea[['Country','Date','Variable','National']]\n",
    "sl_table = sl[['Country','Date','Variable','National']]\n",
    "# Concatenate Tables on Rows\n",
    "merged_table = pd.concat([liberia_table, guinea_table, sl_table], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Reformat Dates\n",
    "merged_table['Date'] = merged_table['Date'].apply(parse)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Method\n",
    "The method choosen relies on creating two additional columns that will contain the number of real death and new cases "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Set country and date as indexes\n",
    "merged_table = merged_table.set_index(['Country','Date'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Exploratory Search for Death'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''Exploratory Search for Death'''\n",
    "# pattern1 = 'death'\n",
    "# '''\n",
    "# HWC: stands for Human wild life conflict,\n",
    "# cum: prbably for cumulative,\n",
    "# CTE, ETC: centre traitement Ebola\n",
    "# '''\n",
    "# Get all deaths modalities in a data frame\n",
    "# pd.Series(pd.unique(merged_table.Variable[merged_table.Variable.str.contains(pattern1)]))\n",
    "\n",
    "# # Type of Deaths\n",
    "# (merged_table[merged_table.Variable.str.contains(pattern1)])\n",
    "\n",
    "# # Type of death in each country \n",
    "# country = 'Sierra Leone'\n",
    "# merged_table.loc[country][merged_table.loc[country].Variable.str.contains(pattern1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Search Results for New Deaths\n",
    "* Liberia = Newly reported deaths, Newly Reported deaths in HCW\n",
    "* Guinea = New deaths registered today, New deaths registered, New deaths registered among health workers\n",
    "* Sierra Leone = death_suspected, death_probable, death_confirmed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculations for Total New Deaths\n",
    "* Liberia = SUM(Newly reported deaths, Newly Reported deaths in HCW)\n",
    "* Guinea: SUM(New deaths registered, New deaths registered,New deaths registered among health workers)\n",
    "* Sierra Leone: SUM(death_suspected[j+1] - death_suspected[j], death_probable[j+1] - death_probable[j], death_confirmed[j+1] - death_confirmed[j])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Choice of real deaths, identify death to take into accout'''\n",
    "deathperday = ['Newly Reported deaths in HCW', 'Newly reported deaths', 'New deaths registered today', 'New deaths registered', 'New deaths registered among health workers', 'death_suspected', 'death_probable', 'death_confirmed']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Exploratory Search for New Cases'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''Exploratory Search for New Cases'''\n",
    "# pattern2 ='ew'\n",
    "\n",
    "# Create a series with cases\n",
    "# cases = pd.Series(pd.unique(merged_table.Variable[merged_table.Variable.str.contains(pattern2)]))\n",
    "# Create a serie with cases that are not deaths nor non cases\n",
    "# cases[cases.str.contains('death|noncase')==False] \n",
    "\n",
    "# Type of of cases in one country \n",
    "# country = 'Sierra Leone'\n",
    "# merged_table.loc[country][merged_table.loc[country].Variable.str.contains(pattern2)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculations for New Cases\n",
    "* Liberia = SUM(New Case/s (Suspected), New Case/s (Probable), New case/s (confirmed))\n",
    "* Guinea = SUM(New cases of confirmed among health workers, Total new cases registered so far)\n",
    "* Sierra Leone = SUM(new_confirmed, new_probable, new_suspected)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Choice of new cases, identify cases to take into accout'''\n",
    "caseperday = ['New Case/s (Suspected)' ,'New Case/s (Probable)', 'New case/s (confirmed)', 'Total new cases registered so far', 'New cases of confirmed among health workers', 'new_confirmed', 'new_probable', 'new_suspected']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Correct Cumalitive Values for Deaths in Sierra Leone"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Suspected Deaths avg per Day: 1.6105263157894736\n",
      "Probable Deaths avg per Day: 1.831578947368421\n",
      "Confirmed Deaths avg per Day: 15.2\n"
     ]
    }
   ],
   "source": [
    "correction = ['death_suspected','death_probable','death_confirmed']\n",
    "cum_table = merged_table[merged_table.Variable.isin(correction)]\n",
    "dates = cum_table.index.tolist()\n",
    "cum_table.National.fillna(0, inplace=True)\n",
    "\n",
    "prev_index = cum_table.index[0]\n",
    "prev_sus_cum = int(cum_table.loc[prev_index].National[cum_table.loc[prev_index].Variable.isin([correction[0]])].values[0])\n",
    "prev_prob_cum = int(cum_table.loc[prev_index].National[cum_table.loc[prev_index].Variable.isin([correction[1]])].values[0])\n",
    "prev_confirm_cum = int(cum_table.loc[prev_index].National[cum_table.loc[prev_index].Variable.isin([correction[2]])].values[0])\n",
    "\n",
    "date = 3\n",
    "sus_total = 0\n",
    "prob_total = 0\n",
    "confirm_total = 0\n",
    "\n",
    "days_sus = 0\n",
    "days_prob = 0\n",
    "days_confirm = 0\n",
    "\n",
    "while date < len(dates):\n",
    "    curr_index = cum_table.index[date]\n",
    "    curr_sus_cum = int(cum_table.loc[curr_index].National[cum_table.loc[curr_index].Variable.isin([correction[0]])].values[0])\n",
    "    curr_prob_cum = int(cum_table.loc[curr_index].National[cum_table.loc[curr_index].Variable.isin([correction[1]])].values[0])\n",
    "    curr_confirm_cum = int(cum_table.loc[curr_index].National[cum_table.loc[curr_index].Variable.isin([correction[2]])].values[0])\n",
    "    \n",
    "    if(curr_sus_cum != 0):\n",
    "        curr_sus = curr_sus_cum - prev_sus_cum\n",
    "        sus_total += curr_sus\n",
    "        prev_sus_cum = curr_sus_cum\n",
    "        days_sus += 1\n",
    "    \n",
    "    if(curr_prob_cum != 0):\n",
    "        curr_prob = curr_prob_cum - prev_prob_cum\n",
    "        prob_total += curr_prob\n",
    "        prev_prob_cum = curr_prob_cum\n",
    "        days_prob += 1\n",
    "        \n",
    "    if(curr_confirm_cum != 0):\n",
    "        curr_confirm = curr_confirm_cum - prev_confirm_cum\n",
    "        confirm_total += curr_confirm\n",
    "        prev_confirm_cum = curr_confirm_cum\n",
    "        days_confirm += 1\n",
    "    \n",
    "    date += 3\n",
    "\n",
    "print('Suspected Deaths avg per Day: ' + str(sus_total/float(days_sus)))\n",
    "print('Probable Deaths avg per Day: ' + str(prob_total/float(days_prob)))\n",
    "print('Confirmed Deaths avg per Day: ' + str(confirm_total/float(days_confirm)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Create a columns with the deaths and one with the new cases and remove Nas\n",
    "merged_table['deaths_per_day'] = np.where(merged_table.Variable.isin(deathperday), merged_table['National'], 0)\n",
    "merged_table['cases_per_day'] = np.where(merged_table.Variable.isin(caseperday), merged_table['National'], 0)\n",
    "merged_table.deaths_per_day.fillna(0, inplace=True)\n",
    "merged_table.cases_per_day.fillna(0, inplace=True)\n",
    "merged_table['deaths_per_day'] = pd.to_numeric(merged_table.deaths_per_day)\n",
    "merged_table['cases_per_day'] = pd.to_numeric(merged_table.cases_per_day)\n",
    "\n",
    "                                     \n",
    "liberia_deaths = merged_table.loc['Liberia'][merged_table.loc['Liberia'].Variable.isin(deathperday)]\n",
    "liberia_new_cases = merged_table.loc['Liberia'][merged_table.loc['Liberia'].Variable.isin(caseperday)]\n",
    "\n",
    "guinea_deaths = merged_table.loc['Guinea'][merged_table.loc['Guinea'].Variable.isin(deathperday)]\n",
    "guinea_new_cases = merged_table.loc['Guinea'][merged_table.loc['Guinea'].Variable.isin(caseperday)]\n",
    "\n",
    "sl_deaths_per_day = (sus_total/float(days_sus)) + (prob_total/float(days_prob)) + (confirm_total/float(days_confirm))\n",
    "sl_new_cases = merged_table.loc['Sierra Leone'][merged_table.loc['Sierra Leone'].Variable.isin(caseperday)]\n",
    "\n",
    "deaths_per_day = [guinea_deaths.deaths_per_day.agg(np.mean), liberia_deaths.deaths_per_day.agg(np.mean), sl_deaths_per_day]\n",
    "new_cases_per_day = [guinea_new_cases.cases_per_day.agg(np.mean), liberia_new_cases.cases_per_day.agg(np.mean), sl_new_cases.cases_per_day.agg(np.mean)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Deaths per Day</th>\n",
       "      <th>New Cases per Day</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Guinea</th>\n",
       "      <td>2.069767</td>\n",
       "      <td>11.302326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Liberia</th>\n",
       "      <td>10.355000</td>\n",
       "      <td>167.146667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sierra Leone</th>\n",
       "      <td>18.642105</td>\n",
       "      <td>17.650485</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Deaths per Day  New Cases per Day\n",
       "Guinea              2.069767          11.302326\n",
       "Liberia            10.355000         167.146667\n",
       "Sierra Leone       18.642105          17.650485"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Countries = ['Guinea', 'Liberia', 'Sierra Leone']\n",
    "Summary = pd.DataFrame(index = Countries)\n",
    "Summary['Deaths per Day'] = deaths_per_day\n",
    "Summary['New Cases per Day'] = new_cases_per_day\n",
    "Summary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2. RNA Sequences\n",
    "\n",
    "In the `DATA_FOLDER/microbiome` subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10<sup>th</sup> file that describes the content of each. \n",
    "\n",
    "Use pandas to import the first 9 spreadsheets into a single `DataFrame`.\n",
    "Then, add the metadata information from the 10<sup>th</sup> spreadsheet as columns in the combined `DataFrame`.\n",
    "Make sure that the final `DataFrame` has a unique index and all the `NaN` values have been replaced by the tag `unknown`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ORGANISM</th>\n",
       "      <th>COUNT</th>\n",
       "      <th>BARCODE</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>SAMPLE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>7</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>2</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Sulfoloba...</td>\n",
       "      <td>3</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Thermopro...</td>\n",
       "      <td>3</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...</td>\n",
       "      <td>7</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Archaeoglobi Archaeogl...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Archaeoglobi Archaeogl...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Halobacteria Halobacte...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Halobacteria Halobacte...</td>\n",
       "      <td>4</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Halobacteria Halobacte...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Halobacteria Halobacte...</td>\n",
       "      <td>2</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Halobacteria Halobacte...</td>\n",
       "      <td>4</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Methanococci Methanoco...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Methanopyri Methanopyr...</td>\n",
       "      <td>12</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Thermoplasmata Thermop...</td>\n",
       "      <td>2</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>2</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>2</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>53</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>2</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>3</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>99</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>17</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>2</td>\n",
       "      <td>MID1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2366</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>10</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2367</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2368</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2369</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2370</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>12</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2371</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>4</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2372</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>805</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2373</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>862</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2374</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>12</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2375</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>3534</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2376</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>4</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2377</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2378</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>25</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>2</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2380</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>3</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2381</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>2</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2382</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2383</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>3</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2384</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>2</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2385</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2386</th>\n",
       "      <td>Bacteria \"Synergistetes\" Synergistia Synergist...</td>\n",
       "      <td>3</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2387</th>\n",
       "      <td>Bacteria \"Tenericutes\" Mollicutes Anaeroplasma...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2388</th>\n",
       "      <td>Bacteria \"Tenericutes\" Mollicutes Haloplasmata...</td>\n",
       "      <td>11</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2389</th>\n",
       "      <td>Bacteria \"Thermodesulfobacteria\" Thermodesulfo...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2390</th>\n",
       "      <td>Bacteria \"Thermodesulfobacteria\" Thermodesulfo...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2391</th>\n",
       "      <td>Bacteria \"Thermotogae\" Thermotogae Thermotogal...</td>\n",
       "      <td>43</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2392</th>\n",
       "      <td>Bacteria \"Thermotogae\" Thermotogae Thermotogal...</td>\n",
       "      <td>16</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2393</th>\n",
       "      <td>Bacteria \"Verrucomicrobia\" Opitutae Opitutales...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2394</th>\n",
       "      <td>Bacteria Cyanobacteria Cyanobacteria  Chloropl...</td>\n",
       "      <td>1</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2395</th>\n",
       "      <td>Bacteria Cyanobacteria Cyanobacteria  Chloropl...</td>\n",
       "      <td>10</td>\n",
       "      <td>MID9</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2396 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               ORGANISM  COUNT BARCODE  \\\n",
       "0     Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...      7    MID1   \n",
       "1     Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...      2    MID1   \n",
       "2     Archaea \"Crenarchaeota\" Thermoprotei Sulfoloba...      3    MID1   \n",
       "3     Archaea \"Crenarchaeota\" Thermoprotei Thermopro...      3    MID1   \n",
       "4     Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...      7    MID1   \n",
       "5     Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...      1    MID1   \n",
       "6     Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...      1    MID1   \n",
       "7     Archaea \"Euryarchaeota\" Archaeoglobi Archaeogl...      1    MID1   \n",
       "8     Archaea \"Euryarchaeota\" Archaeoglobi Archaeogl...      1    MID1   \n",
       "9     Archaea \"Euryarchaeota\" Halobacteria Halobacte...      1    MID1   \n",
       "10    Archaea \"Euryarchaeota\" Halobacteria Halobacte...      4    MID1   \n",
       "11    Archaea \"Euryarchaeota\" Halobacteria Halobacte...      1    MID1   \n",
       "12    Archaea \"Euryarchaeota\" Halobacteria Halobacte...      2    MID1   \n",
       "13    Archaea \"Euryarchaeota\" Halobacteria Halobacte...      4    MID1   \n",
       "14    Archaea \"Euryarchaeota\" Methanococci Methanoco...      1    MID1   \n",
       "15    Archaea \"Euryarchaeota\" Methanopyri Methanopyr...     12    MID1   \n",
       "16    Archaea \"Euryarchaeota\" Thermoplasmata Thermop...      2    MID1   \n",
       "17    Bacteria \"Actinobacteria\" Actinobacteria Actin...      2    MID1   \n",
       "18    Bacteria \"Actinobacteria\" Actinobacteria Actin...      1    MID1   \n",
       "19    Bacteria \"Actinobacteria\" Actinobacteria Actin...      1    MID1   \n",
       "20    Bacteria \"Actinobacteria\" Actinobacteria Actin...      1    MID1   \n",
       "21    Bacteria \"Actinobacteria\" Actinobacteria Actin...      2    MID1   \n",
       "22    Bacteria \"Actinobacteria\" Actinobacteria Actin...     53    MID1   \n",
       "23    Bacteria \"Actinobacteria\" Actinobacteria Actin...      1    MID1   \n",
       "24    Bacteria \"Actinobacteria\" Actinobacteria Actin...      2    MID1   \n",
       "25    Bacteria \"Actinobacteria\" Actinobacteria Actin...      3    MID1   \n",
       "26    Bacteria \"Actinobacteria\" Actinobacteria Actin...      1    MID1   \n",
       "27    Bacteria \"Actinobacteria\" Actinobacteria Actin...     99    MID1   \n",
       "28    Bacteria \"Actinobacteria\" Actinobacteria Actin...     17    MID1   \n",
       "29    Bacteria \"Actinobacteria\" Actinobacteria Actin...      2    MID1   \n",
       "...                                                 ...    ...     ...   \n",
       "2366  Bacteria \"Proteobacteria\" Gammaproteobacteria ...     10    MID9   \n",
       "2367  Bacteria \"Proteobacteria\" Gammaproteobacteria ...      1    MID9   \n",
       "2368  Bacteria \"Proteobacteria\" Gammaproteobacteria ...      1    MID9   \n",
       "2369  Bacteria \"Proteobacteria\" Gammaproteobacteria ...      1    MID9   \n",
       "2370  Bacteria \"Proteobacteria\" Gammaproteobacteria ...     12    MID9   \n",
       "2371  Bacteria \"Proteobacteria\" Gammaproteobacteria ...      4    MID9   \n",
       "2372  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    805    MID9   \n",
       "2373  Bacteria \"Proteobacteria\" Gammaproteobacteria ...    862    MID9   \n",
       "2374  Bacteria \"Proteobacteria\" Gammaproteobacteria ...     12    MID9   \n",
       "2375  Bacteria \"Proteobacteria\" Gammaproteobacteria ...   3534    MID9   \n",
       "2376  Bacteria \"Proteobacteria\" Gammaproteobacteria ...      4    MID9   \n",
       "2377  Bacteria \"Proteobacteria\" Gammaproteobacteria ...      1    MID9   \n",
       "2378  Bacteria \"Proteobacteria\" Gammaproteobacteria ...     25    MID9   \n",
       "2379  Bacteria \"Proteobacteria\" Gammaproteobacteria ...      2    MID9   \n",
       "2380  Bacteria \"Proteobacteria\" Gammaproteobacteria ...      3    MID9   \n",
       "2381  Bacteria \"Proteobacteria\" Gammaproteobacteria ...      2    MID9   \n",
       "2382  Bacteria \"Proteobacteria\" Gammaproteobacteria ...      1    MID9   \n",
       "2383  Bacteria \"Proteobacteria\" Gammaproteobacteria ...      3    MID9   \n",
       "2384  Bacteria \"Proteobacteria\" Gammaproteobacteria ...      2    MID9   \n",
       "2385  Bacteria \"Proteobacteria\" Gammaproteobacteria ...      1    MID9   \n",
       "2386  Bacteria \"Synergistetes\" Synergistia Synergist...      3    MID9   \n",
       "2387  Bacteria \"Tenericutes\" Mollicutes Anaeroplasma...      1    MID9   \n",
       "2388  Bacteria \"Tenericutes\" Mollicutes Haloplasmata...     11    MID9   \n",
       "2389  Bacteria \"Thermodesulfobacteria\" Thermodesulfo...      1    MID9   \n",
       "2390  Bacteria \"Thermodesulfobacteria\" Thermodesulfo...      1    MID9   \n",
       "2391  Bacteria \"Thermotogae\" Thermotogae Thermotogal...     43    MID9   \n",
       "2392  Bacteria \"Thermotogae\" Thermotogae Thermotogal...     16    MID9   \n",
       "2393  Bacteria \"Verrucomicrobia\" Opitutae Opitutales...      1    MID9   \n",
       "2394  Bacteria Cyanobacteria Cyanobacteria  Chloropl...      1    MID9   \n",
       "2395  Bacteria Cyanobacteria Cyanobacteria  Chloropl...     10    MID9   \n",
       "\n",
       "                   GROUP   SAMPLE  \n",
       "0     EXTRACTION CONTROL  unknown  \n",
       "1     EXTRACTION CONTROL  unknown  \n",
       "2     EXTRACTION CONTROL  unknown  \n",
       "3     EXTRACTION CONTROL  unknown  \n",
       "4     EXTRACTION CONTROL  unknown  \n",
       "5     EXTRACTION CONTROL  unknown  \n",
       "6     EXTRACTION CONTROL  unknown  \n",
       "7     EXTRACTION CONTROL  unknown  \n",
       "8     EXTRACTION CONTROL  unknown  \n",
       "9     EXTRACTION CONTROL  unknown  \n",
       "10    EXTRACTION CONTROL  unknown  \n",
       "11    EXTRACTION CONTROL  unknown  \n",
       "12    EXTRACTION CONTROL  unknown  \n",
       "13    EXTRACTION CONTROL  unknown  \n",
       "14    EXTRACTION CONTROL  unknown  \n",
       "15    EXTRACTION CONTROL  unknown  \n",
       "16    EXTRACTION CONTROL  unknown  \n",
       "17    EXTRACTION CONTROL  unknown  \n",
       "18    EXTRACTION CONTROL  unknown  \n",
       "19    EXTRACTION CONTROL  unknown  \n",
       "20    EXTRACTION CONTROL  unknown  \n",
       "21    EXTRACTION CONTROL  unknown  \n",
       "22    EXTRACTION CONTROL  unknown  \n",
       "23    EXTRACTION CONTROL  unknown  \n",
       "24    EXTRACTION CONTROL  unknown  \n",
       "25    EXTRACTION CONTROL  unknown  \n",
       "26    EXTRACTION CONTROL  unknown  \n",
       "27    EXTRACTION CONTROL  unknown  \n",
       "28    EXTRACTION CONTROL  unknown  \n",
       "29    EXTRACTION CONTROL  unknown  \n",
       "...                  ...      ...  \n",
       "2366           Control 2    stool  \n",
       "2367           Control 2    stool  \n",
       "2368           Control 2    stool  \n",
       "2369           Control 2    stool  \n",
       "2370           Control 2    stool  \n",
       "2371           Control 2    stool  \n",
       "2372           Control 2    stool  \n",
       "2373           Control 2    stool  \n",
       "2374           Control 2    stool  \n",
       "2375           Control 2    stool  \n",
       "2376           Control 2    stool  \n",
       "2377           Control 2    stool  \n",
       "2378           Control 2    stool  \n",
       "2379           Control 2    stool  \n",
       "2380           Control 2    stool  \n",
       "2381           Control 2    stool  \n",
       "2382           Control 2    stool  \n",
       "2383           Control 2    stool  \n",
       "2384           Control 2    stool  \n",
       "2385           Control 2    stool  \n",
       "2386           Control 2    stool  \n",
       "2387           Control 2    stool  \n",
       "2388           Control 2    stool  \n",
       "2389           Control 2    stool  \n",
       "2390           Control 2    stool  \n",
       "2391           Control 2    stool  \n",
       "2392           Control 2    stool  \n",
       "2393           Control 2    stool  \n",
       "2394           Control 2    stool  \n",
       "2395           Control 2    stool  \n",
       "\n",
       "[2396 rows x 5 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the metadata file\n",
    "metadata_files = glob.glob(DATA_FOLDER + \"/microbiome\" + \"/metadata.xls\")\n",
    "metadata = pd.pandas.read_excel(metadata_files[0])\n",
    "\n",
    "# Read all the data files\n",
    "data_files = glob.glob(DATA_FOLDER + \"/microbiome\" + \"/MID*.xls\")\n",
    "dfs = []\n",
    "\n",
    "# Join individual data file with metadata\n",
    "for data in data_files:\n",
    "    df = pd.pandas.read_excel(data, header=None, names=[\"ORGANISM\", \"COUNT\"])\n",
    "    df[\"BARCODE\"] = data[-8:-4]\n",
    "    df_full = df.join(metadata.set_index(\"BARCODE\"), on=\"BARCODE\")\n",
    "    dfs.append(df_full)\n",
    "\n",
    "# Concatenate all data files\n",
    "result = pd.concat(dfs)\n",
    "\n",
    "# Reindex and replace NA\n",
    "result.index = np.arange(len(result))\n",
    "result.fillna(\"unknown\", inplace=True)\n",
    "\n",
    "# Print result and check if indices are unique\n",
    "result.index.is_unique\n",
    "result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3. Class War in Titanic\n",
    "\n",
    "Use pandas to import the data file `Data/titanic.xls`. It contains data on all the passengers that travelled on the Titanic."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each of the following questions state clearly your assumptions and discuss your findings:\n",
    "1. Describe the *type* and the *value range* of each attribute. Indicate and transform the attributes that can be `Categorical`. \n",
    "2. Plot histograms for the *travel class*, *embarkation port*, *sex* and *age* attributes. For the latter one, use *discrete decade intervals*. \n",
    "3. Calculate the proportion of passengers by *cabin floor*. Present your results in a *pie chart*.\n",
    "4. For each *travel class*, calculate the proportion of the passengers that survived. Present your results in *pie charts*.\n",
    "5. Calculate the proportion of the passengers that survived by *travel class* and *sex*. Present your results in *a single histogram*.\n",
    "6. Create 2 equally populated *age categories* and calculate survival proportions by *age category*, *travel class* and *sex*. Present your results in a `DataFrame` with unique index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#Import files\n",
    "from IPython.core.display import HTML\n",
    "HTML(filename=DATA_FOLDER+'/titanic.html')\n",
    "\n",
    "\n",
    "HTML(filename=DATA_FOLDER+'/titanic.html')\n",
    "\n",
    "#Import from xls\n",
    "my_titanic=pd.pandas.read_excel(DATA_FOLDER+'/titanic.xls')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "HTML(filename=DATA_FOLDER+'/titanic.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''Fill NaN with a tag'''\n",
    "my_titanic.cabin.fillna('unknown', inplace=True)\n",
    "my_titanic.age.fillna(int(-1), inplace=True) # all the unknown values will get a numeric value <0 for plotting purposes\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Correct Cabin Syntax'''\n",
    "foo=my_titanic.cabin[my_titanic.cabin.str.contains('F ').fillna(False)]\n",
    "for i in foo.index:\n",
    "    print('replace {} with {}'.format (my_titanic.cabin.iloc[i], my_titanic.cabin.iloc[i][2:]))\n",
    "    my_titanic.cabin.iloc[i]=my_titanic.cabin.iloc[i][2:]\n",
    "    #the string contains function has a problem with na=True argument \n",
    "foo=my_titanic.cabin[my_titanic.cabin.str.contains('T').fillna(False)]\n",
    "for i in foo.index:\n",
    "    print('replace {} with {}'.format (my_titanic.cabin.iloc[i], 'unknown'))\n",
    "    my_titanic.cabin.iloc[i]='unknown'\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "'''Create a new Column for the Cabin floors '''\n",
    "my_titanic['cabin_floor']=0\n",
    "for i in my_titanic['cabin'].index:\n",
    "    if my_titanic.cabin.iloc[i]=='unknown':\n",
    "        my_titanic['cabin_floor'][i]=my_titanic['cabin'][i]\n",
    "    else: \n",
    "        my_titanic['cabin_floor'][i]=my_titanic['cabin'][i][0]\n",
    "my_titanic['cabin_floor']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "''' Describe the *type* and the *value range* of each attribute. \n",
    "Indicate and transform the attributes that can be `Categorical`'''\n",
    "\n",
    "for i in range(my_titanic.shape[1]):\n",
    "    print(\"Attribute '{}' is of type {}\".format(my_titanic.columns[i],my_titanic[my_titanic.columns[i]].dtype ),end=\" \")\n",
    "          #my_titanic[my_titanic.columns[i]].head())\n",
    "    if my_titanic[my_titanic.columns[i]].dtype==(int or float):\n",
    "        print(\"its Value Range is {} to {}\"  \\\n",
    "              .format(my_titanic[my_titanic.columns[i]].min(),format(my_titanic[my_titanic.columns[i]].max() )))\n",
    "    elif my_titanic[my_titanic.columns[i]].dtype==object:\n",
    "        print(\"its has {} unique categories\"  \\\n",
    "              .format(my_titanic[my_titanic.columns[i]].describe()[1] ))\n",
    "        my_titanic[my_titanic.columns[i]] = my_titanic[my_titanic.columns[i]].astype('category') # categorical data \n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Question 1: \n",
    "\n",
    "Attribute 'pclass' is of type int64 its Value Range is 1 to 3\n",
    "Attribute 'survived' is of type int64 its Value Range is 0 to 1\n",
    "Attribute 'name' is of type object its has 1307 unique categories\n",
    "Attribute 'sex' is of type object its has 2 unique categories\n",
    "Attribute 'age' is of type float64 \n",
    "Attribute 'sibsp' is of type int64 its Value Range is 0 to 8\n",
    "Attribute 'parch' is of type int64 its Value Range is 0 to 9\n",
    "Attribute 'ticket' is of type object its has 939 unique categories\n",
    "Attribute 'fare' is of type float64 \n",
    "Attribute 'cabin' is of type object its has 186 unique categories\n",
    "Attribute 'embarked' is of type object its has 3 unique categories\n",
    "Attribute 'boat' is of type object its has 28 unique categories\n",
    "Attribute 'body' is of type float64\n",
    "Attribute 'home.dest' is of type object its has 369 unique categories\n",
    "Attribute 'cabin_floor' is of type object its has 8 unique categories # Obtained from \"Cabin\" \n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "'''Plot histograms for the *travel class*, *embarkation port*,\n",
    "*sex* and *age* attributes. For the latter one, use *discrete decade intervals*.'''\n",
    "\n",
    "ax=my_titanic.pclass.hist(normed=True)\n",
    "ax.set_ylabel('Number of Passenger')\n",
    "ax.set_xlabel('Passenger Class')\n",
    "plt.show()\n",
    "\n",
    "new_age=pd.cut(my_titanic.age, [-2,0,10,20,30,40,50,60,70,80,90]).value_counts(sort=False)\n",
    "bx=new_age.plot.bar()\n",
    "bx.set_ylabel('Number of Passenger')\n",
    "bx.set_xlabel('Age Range')\n",
    "bx.set_xticklabels(['unknown','0 to 10','10 to 20','20 to 30',\\\n",
    "'30 to 40','40 to 50','50 to 60','60 to 70','70 to 80','80 to 90','90 to 100'])\n",
    "#ax1.set_xticklabels=new_age\n",
    "plt.show()\n",
    "\n",
    "cx=my_titanic.embarked.value_counts().plot.bar()\n",
    "cx.set_ylabel('Number of Passenger')\n",
    "cx.set_xlabel('Embarquement port')\n",
    "#ax1.set_xticklabels([c[1:-1].replace(\",\",\" to\") for c in new_age.cat.categories])\n",
    "plt.show()\n",
    "\n",
    "\n",
    "dx=my_titanic.sex.value_counts().plot.bar()\n",
    "dx.set_ylabel('Number of Passenger')\n",
    "dx.set_xlabel('Sex')\n",
    "plt.show()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Question 2: \n",
    "Plot histograms for the travel class, embarkation port, sex and age attributes. For the latter one, use discrete decade intervals.\n",
    "\n",
    "Discussion: Half of the passengers were in the 3rd class. The second half was equally split between class 1 and 2 \n",
    "Most of the passenger were under 30, and more than half is less than 40 years old. We make the assumption that the ages with decimals <0 are correct because people travelled wit their babies \n",
    "Most of the passengers embarked from Southampton, the rest emarked from Cherbourg and Queenstown and 'unknown' ports\n",
    "There was almost twice as much males than females on the titanic. \n",
    "\t \n",
    "\t\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "'''Calculate the proportion of passengers by *cabin floor*.'''\n",
    "dx=my_titanic.cabin_floor.value_counts().plot.pie()\n",
    "dx.set_ylabel('Proportion of Passenger')\n",
    "dx.set_xlabel('Cabin floor ')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Question 3: \n",
    "Here I have assumed that the first letter of the Cabin field indicated the floor of the boat. Therefore I have added a new field to my data table prior to Categorical values transformation. For the cabin argument we make a simplification choosing only the first letter of the cabin string. The cabin T and NaN are replaced by an unknown field. For F E57 types only the second letter is used \n",
    "\n",
    "We observe that most passengers were concentrated on floors B, C, D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "'''For each *travel class*, calculate the proportion of the passengers that survived. '''\n",
    "\n",
    "\n",
    "#Group by Class\n",
    "tmp=my_titanic.groupby(['pclass'])['survived'].value_counts()\n",
    "tmp=tmp.unstack()\n",
    "tmp.columns=['Death','Survival']\n",
    "tmp['Sum_pergroup']=tmp.sum(axis=1)\n",
    "tmp2=tmp.copy()\n",
    "tmp['Death']=round(tmp.Death/tmp.Sum_pergroup*100)\n",
    "tmp['Survival']=round(tmp.Survival/tmp.Sum_pergroup*100)\n",
    "#tmp['Death','Survival'] =[tmp.Death/tmp.Sum_pergroup,tmp.Survival/tmp.Sum_pergroup]\n",
    "print(tmp)\n",
    "print(tmp2)\n",
    "\n",
    "#Plot Version 2\n",
    "forplot=tmp.iloc[:,0:2].plot.pie(subplots=True)\n",
    "#forplot.set_ylabel('Percent')\n",
    "#forplot.set_xlabel('Travel Class')\n",
    "plt.show()\n",
    "\n",
    "#Plot Version 1\n",
    "for i in range(3):\n",
    "    forplot1=tmp.iloc[i,0:2].plot.pie()\n",
    "    forplot1.set_ylabel('Travel Class {}'.format(i+1))\n",
    "    plt.show()\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " Question 4: Here the proportions of survivors are 62, 43, 26 percent for travel classes 1,2,3 respectively "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Calculate the proportion of the passengers that survived by *travel class* and *sex*.'''\n",
    "tmp=my_titanic.groupby(['pclass','sex'])['survived'].value_counts()\n",
    "tmp=tmp.unstack()\n",
    "tmp.columns=['Death','Survival']\n",
    "                #print(tmp.loc[1,'female']['Death'])\n",
    "tmp['Sum_pergroup']=tmp.sum(axis=1)\n",
    "tmp2=tmp.copy()\n",
    "tmp['Death']=round(tmp.Death/tmp.Sum_pergroup*100)\n",
    "tmp['Survival']=round(tmp.Survival/tmp.Sum_pergroup*100)\n",
    "#tmp['Death','Survival'] =[tmp.Death/tmp.Sum_pergroup,tmp.Survival/tmp.Sum_pergroup]\n",
    "print(tmp)\n",
    "print(tmp2)\n",
    "\n",
    "\n",
    "forplot=tmp[['Death','Survival']].plot.bar(stacked=True)\n",
    "forplot.set_ylabel('Percent')\n",
    "forplot.set_xlabel('Travel Class, Sex')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Create 2 equally populated *age categories*'''\n",
    "'''Equal Partition'''\n",
    "my_titanic.age.fillna(int(-1), inplace=True) #this leaves some NaN!!\n",
    "# Create a new column of ages with the rounded value and the unknown values\n",
    "my_titanic['new_ages_Class']=0\n",
    "my_titanic['new_ages_Class'][my_titanic.age>=0]=round(my_titanic.age[my_titanic.age>=1])\n",
    "my_titanic['new_ages_Class'][my_titanic.age<0]=my_titanic.age[my_titanic.age<0]\n",
    "tmp=my_titanic.groupby(['new_ages_Class'])['new_ages_Class'].value_counts()\n",
    "\n",
    "'''Look for ideal partition by permuting ages until the partition is exact'''\n",
    "#Total Passenger/ 2 =654.5 persons\n",
    "bar=0 \n",
    "while bar==0:\n",
    "    shuffle_indices = np.random.permutation(len(tmp))\n",
    "    tmp=tmp[shuffle_indices]\n",
    "    for i,vali in enumerate(tmp.cumsum()):\n",
    "        if vali==654 | 655:\n",
    "            partition_ix=i\n",
    "            print('true')\n",
    "            bar=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sum(tmp[0:partition_ix+1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''Partition according to ideal partition'''\n",
    "my_titanic['age.partition']=0\n",
    "left_partition=list(tmp[partition_ix+2:].index.labels[1])\n",
    "right_partition=list(tmp[0:partition_ix+1].index.labels[1])\n",
    "my_titanic['age.partition'][my_titanic['new_ages_Class'].isin(right_partition)]=1      \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sum(my_titanic['age.partition'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Calculate survival proportions by *age category*, *travel class* and *sex*. '''\n",
    "tmp=my_titanic.groupby(['age.partition','pclass','sex'])['survived'].value_counts()\n",
    "tmp=tmp.unstack()\n",
    "tmp.columns=['Death','Survival']\n",
    "tmp['Sum_pergroup']=tmp.sum(axis=1)\n",
    "tmp['Death']=round(tmp.Death/tmp.Sum_pergroup*100)\n",
    "tmp['Survival']=round(tmp.Survival/tmp.Sum_pergroup*100)\n",
    "tmp=tmp.iloc[:,0:2]\n",
    "print(tmp)\n",
    "tmp.index.is_unique #Unique (Multi Index )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your answer here\n",
    "\n",
    "print(html.unescape('&pound;682m'))\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
